apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -2412,6 +2414,38 @@ flush_messages(N, Acc) ->
               {N, Acc}
     end.
 
+
+%%%----------------------------------------------------------------------
+%%% XEP-0016
+%%%----------------------------------------------------------------------
+
+maybe_update_presence(StateData = #state{jid = JID}, NewList) ->
+    {Froms, _Tos, _Pending} = ejabberd_hooks:run_fold(
```

Comment:
Ah, thanks. I hadn't noticed that they were already in the state. I've pushed a fix, as well as tweaking some of the test code a little to avoid a couple of intermittent failures I was getting.